positional_encodings/positional_encodings.py

```diff
@@ -76,7 +76,40 @@ def forward(self, tensor):
         emb[:, :, : self.channels] = emb_x
         emb[:, :, self.channels : 2 * self.channels] = emb_y
 
-        return emb[None, :, :, :orig_ch].repeat(batch_size, 1, 1, 1)
+        return emb[None, :, :, :orig_ch]
+
+class FixedPositionalEncoding2D(nn.Module):
+    def __init__(self, shape, channels):
+        super(FixedPositionalEncoding2D, self).__init__()
+        self.pos_encoder = SinglePositionalEncoding2D(channels)
+        self.pos_encoding = self.pos_encoder(torch.ones(1, shape[0], shape[1], channels))
+        self.device = self.pos_encoding.device
+        self.repeated_pos_encoding = self.pos_encoding.repeat(32, 1, 1, 1)
+
+    def forward(self, tensor):
+        # Ensure encoding on same device
+        if self.device is not tensor.device:
+            self.pos_encoding = self.pos_encoding.to(tensor.device)
+            self.device = self.pos_encoding.device
```

Comment:
Would just doing `self.pos_encoding.to(tensor.device)` not work? from the docs:
https://pytorch.org/docs/stable/generated/torch.Tensor.to.html#torch.Tensor.to

```
If the self Tensor already has the correct torch.dtype and torch.device, then self is returned. Otherwise, the returned tensor is a copy of self with the desired torch.dtype and torch.device.
```

Reply:
You are probably right about this! I don't know which of the two is the better one. I will test if it makes an impact on the speed, if not, I think your solution does indeed look cleaner. 

The problem is that the pos_encoding is not automatically on the same device as the model. I which there is some method which makes this tensor move when you perform: model.to(device). If this is possible I think this would be the best solution. Do you know if this is possible?

Reply:
I am not sure there is something like that :cry:. That would indeed be the best...

Reply:
Yeah, it feels kind of dirty. Also, it could result into and error atm if someone would switch from device of the data. This because I left out the device check in the newer version because it affect the speed quite a bit if I remember correctly. Maybe I could add the if statement again because that one was a bit less taxing if I think. But not today.